README.md
# chef-teamcity-cookbook

Gives you the ability to create TeamCity server/agent

## Supported Platforms

* CentOS 6.5
* Redhat 6.5
* Windows (agent only)

## Depends

* java
* git
* mercurial
* subversion

## Attributes

<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['teamcity']['version']</tt></td>
    <td>String</td>
    <td>The version of TeamCity.</td>
    <td><tt>8.1.5</tt></td>
  </tr> 
   <tr>
    <td><tt>['teamcity']['username']</tt></td>
    <td>String</td>
    <td>The username that TeamCity will be running under.</td>
    <td><tt>teamcity</tt></td>
    </tr> 
  <tr>
    <td><tt>['teamcity']['password']</tt></td>
    <td>String</td>
    <td>The password that TeamCity will be running under.</td>
    <td></td>
  </tr> 
  <tr>
    <td><tt>['teamcity']['group']</tt></td>
    <td>String</td>
    <td>The group that TeamCity will be running under.</td>
    <td><tt>teamcity</tt></td>
  </tr>
  <tr>
    <td><tt>['teamcity']['service_name']</tt></td>
    <td>String</td>
    <td>The service name of TeamCity.</td>
    <td><tt>teamcity</tt></td>
  </tr>
  <tr>
    <td><tt>['teamcity']['server']['backup']</tt></td>
    <td>String</td>
    <td>The URI of the TeamCity backup.</td>
    <td><tt></tt></td>
  </tr>
  <tr>
    <td><tt>['teamcity']['server']['database']['username']</tt></td>
    <td>String</td>
    <td>The database user name.</td>
    <td><tt></tt></td>
  </tr>
  <tr>
    <td><tt>['teamcity']['server']['database']['password']</tt></td>
    <td>String</td>
    <td>The database password.</td>
    <td><tt></tt></td>
  </tr>
  <tr>
    <td><tt>['teamcity']['server']['database']['connection_url']</tt></td>
    <td>String</td>
    <td>The JDBC connection URL.</td>
    <td><tt></tt></td>
  </tr>
  <tr>
    <td><tt>['teamcity']['server']['database']['jar']</tt></td>
    <td>String</td>
    <td>The URI of the database JAR file.</td>
    <td><tt></tt></td>
  </tr>
  <tr>
    <td><tt>['teamcity']['agent']['name']</tt></td>
    <td>String</td>
    <td>The agent name.</td>
    <td><tt></tt></td>
  </tr>
    <tr>
    <td><tt>['teamcity']['agent']['server_uri']</tt></td>
    <td>String</td>
    <td>The URI of the TeamCity server.</td>
    <td><tt></tt></td>
  </tr>
  <tr>
    <td><tt>['teamcity']['agent']['own_address']</tt></td>
    <td>String</td>
    <td>The IP address of the TeamCity agent.</td>
    <td><tt>nil (will be set automatically)</tt></td>
  </tr>
  <tr>
    <td><tt>['teamcity']['agent']['port']</tt></td>
    <td>String</td>
    <td>The The port to use for the teamcity agent.</td>
    <td><tt>9090</tt></td>
  </tr>
  <tr>
    <td><tt>['teamcity']['agent']['authorization_token']</tt></td>
    <td>String</td>
    <td>The authorization token for the agent to be authorized on the teamcity server.</td>
    <td><tt>nil</tt></td>
  </tr>
  <tr>
    <td><tt>['teamcity']['agent']['system_properties']</tt></td>
    <td>String</td>
    <td>Build Script Properties.</td>
    <td><tt></tt></td>
  </tr>
  <tr>
    <td><tt>['teamcity']['agent']['env_properties']</tt></td>
    <td>String</td>
    <td>Environment Variables.</td>
    <td><tt></tt></td>
  </tr>
  <tr>
    <td><tt>['teamcity']['agent']['work_dir']</tt></td>
    <td>String</td>
    <td>The work directory of the teamcity agent.</td>
    <td><tt>windows: C:/teamcity/work || linux: ../work</tt></td>
  </tr>
  <tr>
    <td><tt>['teamcity']['agent']['temp_dir']</tt></td>
    <td>String</td>
    <td>The temp directory of the teamcity agent.</td>
    <td><tt>windows: C:/teamcity/temp || linux: ../temp</tt></td>
  </tr>
  <tr>
    <td><tt>['teamcity']['agent']['system_dir']</tt></td>
    <td>String</td>
    <td>The main directory of the teamcity agent.</td>
    <td><tt>windows: C:/teamcity || linux: ../system</tt></td>
  </tr>
</table>

## Usage

### chef-teamcity::server

Include `chef-teamcity` in your node's `run_list`:

```json
{
  "run_list": [
    "recipe[chef-teamcity::server]"
  ]
}
```

### chef-teamcity::agent

Include `chef-teamcity` in your node's `run_list`:

```json
{
  "run_list": [
    "recipe[chef-teamcity::agent]"
  ]
}
```

## License and Authors

- Author:: Alex Falkowski (<alexrfalkowski@gmail.com>)

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# 0.1.2

- Remove subversion and mercurial

# 0.1.1

* Updated default.rb to work with Chef Client 13

# 0.1.0

* Initial release of chef-teamcity

# 0.2.0

* Fix groups - [7](https://github.com/alexfalkowski/chef-teamcity/pull/7).

# 0.3.0

* Add java recipes - [8](https://github.com/alexfalkowski/chef-teamcity/pull/8).
* Added stove gem with version 3.2.3 - [9](https://github.com/alexfalkowski/chef-teamcity/pull/9).

# 0.4.0

* Don't regenerate the properties file if it already exists. Teamcity stores the authorizationToken there. - [10](https://github.com/alexfalkowski/chef-teamcity/pull/10).

# 0.5.0

* Make sure we don't process the same backup. - [11](https://github.com/alexfalkowski/chef-teamcity/pull/11).

# 0.6.0

* Make sure we specify the creation of a home directory. - [14](https://github.com/alexfalkowski/chef-teamcity/pull/14).
* Added git, mercurial and subversion.. - [15](https://github.com/alexfalkowski/chef-teamcity/pull/15).

# 0.7.0

* Removed duplication between agent and server. - [16](https://github.com/alexfalkowski/chef-teamcity/pull/16)
* Added support for Teamcity 9.0 - [16](https://github.com/alexfalkowski/chef-teamcity/pull/16)


recipes/server.rb
#
# Cookbook Name:: chef-teamcity
# Recipe:: server
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

TEAMCITY_VERSION = node['teamcity']['version'].freeze
TEAMCITY_USERNAME = node['teamcity']['username'].freeze
TEAMCITY_SERVICE_NAME = node['teamcity']['service_name'].freeze
TEAMCITY_GROUP = node['teamcity']['group'].freeze
TEAMCITY_HOME_PATH = "/home/#{TEAMCITY_USERNAME}".freeze
TEAMCITY_PATH = "/opt/TeamCity-#{TEAMCITY_VERSION}".freeze
TEAMCITY_INIT_LOCATION = "/etc/init.d/#{TEAMCITY_SERVICE_NAME}".freeze
TEAMCITY_EXECUTABLE_MODE = 0755
TEAMCITY_READ_MODE = 0644

TEAMCITY_SRC_PATH = "#{TEAMCITY_PATH}.tar.gz".freeze
TEAMCITY_PID_FILE = "#{TEAMCITY_PATH}/logs/#{TEAMCITY_SERVICE_NAME}.pid".freeze
TEAMCITY_DB_USERNAME = node['teamcity']['server']['database']['username'].freeze
TEAMCITY_DB_PASSWORD = node['teamcity']['server']['database']['password'].freeze
TEAMCITY_DB_CONNECTION_URL = node['teamcity']['server']['database']['connection_url'].freeze
TEAMCITY_SERVER_EXECUTABLE = "#{TEAMCITY_PATH}/bin/teamcity-server.sh".freeze
TEAMCITY_BIN_PATH = "#{TEAMCITY_PATH}/bin".freeze
TEAMCITY_DATA_PATH = "#{TEAMCITY_PATH}/.BuildServer".freeze
TEAMCITY_LIB_PATH = "#{TEAMCITY_DATA_PATH}/lib".freeze
TEAMCITY_JDBC_PATH = "#{TEAMCITY_LIB_PATH}/jdbc".freeze
TEAMCITY_CONFIG_PATH = "#{TEAMCITY_DATA_PATH}/config".freeze
TEAMCITY_BACKUP_PATH = "#{TEAMCITY_DATA_PATH}/backup".freeze
TEAMCITY_DATABASE_PROPS_NAME = 'database.properties'.freeze
TEAMCITY_DATABASE_PROPS_PATH = "#{TEAMCITY_CONFIG_PATH}/#{TEAMCITY_DATABASE_PROPS_NAME}".freeze
TEAMCITY_JAR_URI = node['teamcity']['server']['database']['jar'].freeze
TEAMCITY_BACKUP_FILE = node['teamcity']['server']['backup']
TEAMCITY_JAR_NAME = ::File.basename(URI.parse(TEAMCITY_JAR_URI).path).freeze

include_recipe 'chef-teamcity::default'

remote_file TEAMCITY_SRC_PATH do
  source "http://download.jetbrains.com/teamcity/TeamCity-#{TEAMCITY_VERSION}.tar.gz"
  owner TEAMCITY_USERNAME
  group TEAMCITY_GROUP
  mode TEAMCITY_READ_MODE
end

bash 'extract_teamcity' do
  cwd '/opt'
  code <<-EOH
    mkdir -p #{TEAMCITY_PATH}
    tar xzf #{TEAMCITY_SRC_PATH} -C #{TEAMCITY_PATH}
    mv #{TEAMCITY_PATH}/*/* #{TEAMCITY_PATH}/
    chown -R #{TEAMCITY_USERNAME}.#{TEAMCITY_GROUP} #{TEAMCITY_PATH}
  EOH
  not_if { ::File.exist?(TEAMCITY_PATH) }
end

paths = [
  TEAMCITY_DATA_PATH,
  TEAMCITY_LIB_PATH,
  TEAMCITY_JDBC_PATH,
  TEAMCITY_CONFIG_PATH,
  TEAMCITY_BACKUP_PATH
]

paths.each do |p|
  directory p do
    owner TEAMCITY_USERNAME
    group TEAMCITY_GROUP
    recursive true
    mode TEAMCITY_EXECUTABLE_MODE
  end
end

remote_file "#{TEAMCITY_JDBC_PATH}/#{TEAMCITY_JAR_NAME}" do
  source TEAMCITY_JAR_URI
  owner TEAMCITY_USERNAME
  group TEAMCITY_GROUP
  mode TEAMCITY_READ_MODE
end

if TEAMCITY_BACKUP_FILE
  backup_file = ::File.basename(URI.parse(TEAMCITY_BACKUP_FILE).path).freeze
  processed_backup_file = File.basename(backup_file, '.*').freeze
  backup_path = ::File.join(TEAMCITY_BACKUP_PATH, backup_file).freeze
  processed_backup_path = ::File.join(TEAMCITY_BACKUP_PATH, processed_backup_file).freeze
  home_database_props = ::File.join(TEAMCITY_HOME_PATH, TEAMCITY_DATABASE_PROPS_NAME).freeze

  remote_file backup_path do
    source TEAMCITY_BACKUP_FILE
    owner TEAMCITY_USERNAME
    group TEAMCITY_GROUP
    mode TEAMCITY_READ_MODE
    not_if { ::File.exist?(processed_backup_path) }
  end

  template home_database_props do
    source 'database.properties.erb'
    mode TEAMCITY_READ_MODE
    owner TEAMCITY_USERNAME
    group TEAMCITY_GROUP
    variables(
                url: TEAMCITY_DB_CONNECTION_URL,
                username: TEAMCITY_DB_USERNAME,
                password: TEAMCITY_DB_PASSWORD
              )
  end

  bash 'restore' do
    user TEAMCITY_USERNAME
    group TEAMCITY_GROUP
    code <<-EOH
      #{TEAMCITY_BIN_PATH}/maintainDB.sh restore -F #{backup_file} -A #{TEAMCITY_DATA_PATH} -T #{home_database_props}
      rm -f #{backup_path}
      touch #{processed_backup_path}
    EOH
    not_if { ::File.exist?(processed_backup_path) }
  end
end

template TEAMCITY_DATABASE_PROPS_PATH do
  source 'database.properties.erb'
  mode TEAMCITY_READ_MODE
  owner TEAMCITY_USERNAME
  group TEAMCITY_GROUP
  variables(
              url: TEAMCITY_DB_CONNECTION_URL,
              username: TEAMCITY_DB_USERNAME,
              password: TEAMCITY_DB_PASSWORD
            )
  notifies :restart, "service[#{TEAMCITY_SERVICE_NAME}]", :delayed
end

template TEAMCITY_INIT_LOCATION do
  source 'teamcity_server_init.erb'
  mode TEAMCITY_EXECUTABLE_MODE
  owner 'root'
  group 'root'
  variables(
              teamcity_user_name: TEAMCITY_USERNAME,
              teamcity_server_executable: TEAMCITY_SERVER_EXECUTABLE,
              teamcity_data_path: TEAMCITY_DATA_PATH,
              teamcity_pidfile: TEAMCITY_PID_FILE,
              teamcity_service_name: TEAMCITY_SERVICE_NAME
            )
  notifies :restart, "service[#{TEAMCITY_SERVICE_NAME}]", :delayed
end

service TEAMCITY_SERVICE_NAME do
  supports start: true, stop: true, restart: true, status: true
  action [:enable, :start]
end


recipes/default.rb
#
# Cookbook Name:: chef-teamcity
# Recipe:: default
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

TEAMCITY_USERNAME = node['teamcity']['username'].freeze
TEAMCITY_PASSWORD = node['teamcity']['password'].freeze
TEAMCITY_GROUP = node['teamcity']['group'].freeze
TEAMCITY_HOME_PATH = "/home/#{TEAMCITY_USERNAME}".freeze

include_recipe 'java'

include_recipe 'git'

if node['platform'] != 'windows'
  package 'git'

  group TEAMCITY_GROUP

  user TEAMCITY_USERNAME do
    manage_home true
    home TEAMCITY_HOME_PATH
    gid TEAMCITY_GROUP
    shell '/bin/bash'
    password TEAMCITY_PASSWORD
  end
end


recipes/linux_agent.rb
#
# Cookbook Name:: chef-teamcity
# Recipe:: agent-linux
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

TEAMCITY_VERSION = node['teamcity']['version'].freeze
TEAMCITY_USERNAME = node['teamcity']['username'].freeze
TEAMCITY_SERVICE_NAME = node['teamcity']['service_name'].freeze
TEAMCITY_GROUP = node['teamcity']['group'].freeze
TEAMCITY_HOME_PATH = "/home/#{TEAMCITY_USERNAME}".freeze
TEAMCITY_PATH = "/opt/TeamCity-#{TEAMCITY_VERSION}".freeze
TEAMCITY_INIT_LOCATION = "/etc/init.d/#{TEAMCITY_SERVICE_NAME}".freeze
TEAMCITY_EXECUTABLE_MODE = 0755
TEAMCITY_READ_MODE = 0644

TEAMCITY_SRC_PATH = "#{TEAMCITY_PATH}.zip".freeze
TEAMCITY_PID_FILE = "#{TEAMCITY_PATH}/logs/buildAgent.pid".freeze
TEAMCITY_AGENT_NAME = node['teamcity']['agent']['name'].freeze
TEAMCITY_AGENT_SERVER_URI = node['teamcity']['agent']['server_uri'].freeze
TEAMCITY_AGENT_FILE = 'buildAgent.zip'.freeze
TEAMCITY_AGENT_URI = ::URI.join(TEAMCITY_AGENT_SERVER_URI, "update/#{TEAMCITY_AGENT_FILE}").to_s.freeze
TEAMCITY_AGENT_SRC_PATH = ::File.join(TEAMCITY_PATH, TEAMCITY_AGENT_FILE).freeze
TEAMCITY_AGENT_CONFIG_PATH = "#{TEAMCITY_PATH}/conf".freeze
TEAMCITY_AGENT_PROPERTIES = "#{TEAMCITY_AGENT_CONFIG_PATH}/buildAgent.properties".freeze
TEAMCITY_AGENT_EXECUTABLE = "#{TEAMCITY_PATH}/bin/agent.sh".freeze
TEAMCITY_AGENT_WORK_DIR = node['teamcity']['agent']['work_dir'].freeze
TEAMCITY_AGENT_TEMP_DIR = node['teamcity']['agent']['temp_dir'].freeze
TEAMCITY_AGENT_SYSTEM_DIR = node['teamcity']['agent']['system_dir'].freeze
TEAMCITY_AGENT_OWN_ADDRESS = node['teamcity']['agent']['own_address'].freeze
TEAMCITY_AGENT_OWN_PORT = node['teamcity']['agent']['port'].freeze
TEAMCITY_AGENT_AUTH_TOKEN = node['teamcity']['agent']['authorization_token'].freeze
TEAMCITY_AGENT_SYSTEM_PROPERTIES = node['teamcity']['agent']['system_properties'].freeze
TEAMCITY_AGENT_ENV_PROPERTIES = node['teamcity']['agent']['env_properties'].freeze

package 'unzip'

remote_file TEAMCITY_SRC_PATH do
  source TEAMCITY_AGENT_URI
  owner TEAMCITY_USERNAME
  group TEAMCITY_GROUP
  mode TEAMCITY_READ_MODE
  not_if { ::File.exist?(TEAMCITY_PATH) }
end

bash 'extract_teamcity' do
  cwd '/opt'
  code <<-EOH
    unzip #{TEAMCITY_SRC_PATH} -d #{TEAMCITY_PATH}
    chown -R #{TEAMCITY_USERNAME}.#{TEAMCITY_GROUP} #{TEAMCITY_PATH}
    rm -f #{TEAMCITY_SRC_PATH}
  EOH
  not_if { ::File.exist?(TEAMCITY_PATH) }
end

file TEAMCITY_AGENT_EXECUTABLE do
  owner TEAMCITY_USERNAME
  group TEAMCITY_GROUP
  mode TEAMCITY_EXECUTABLE_MODE
end

[TEAMCITY_AGENT_CONFIG_PATH].each do |p|
  directory p do
    owner TEAMCITY_USERNAME
    group TEAMCITY_GROUP
    recursive true
    mode TEAMCITY_EXECUTABLE_MODE
  end
end

template TEAMCITY_AGENT_PROPERTIES do
  source 'buildAgent.properties.erb'
  mode TEAMCITY_READ_MODE
  owner TEAMCITY_USERNAME
  group TEAMCITY_GROUP
  variables(
              server_uri: TEAMCITY_AGENT_SERVER_URI,
              name: TEAMCITY_AGENT_NAME,
              work_dir: TEAMCITY_AGENT_WORK_DIR,
              temp_dir: TEAMCITY_AGENT_TEMP_DIR,
              system_dir: TEAMCITY_AGENT_SYSTEM_DIR,
              own_address: TEAMCITY_AGENT_OWN_ADDRESS,
              own_port: TEAMCITY_AGENT_OWN_PORT,
              authorization_token: TEAMCITY_AGENT_AUTH_TOKEN,
              system_properties: TEAMCITY_AGENT_SYSTEM_PROPERTIES,
              env_properties: TEAMCITY_AGENT_ENV_PROPERTIES
            )
  not_if { ::File.exist?(TEAMCITY_AGENT_PROPERTIES) }
  notifies :restart, "service[#{TEAMCITY_SERVICE_NAME}]", :delayed
end

template TEAMCITY_INIT_LOCATION do
  source 'teamcity_agent_init.erb'
  mode TEAMCITY_EXECUTABLE_MODE
  owner 'root'
  group 'root'
  variables(
              teamcity_user_name: TEAMCITY_USERNAME,
              teamcity_executable: TEAMCITY_AGENT_EXECUTABLE,
              teamcity_pidfile: TEAMCITY_PID_FILE,
              teamcity_service_name: TEAMCITY_SERVICE_NAME
            )
  notifies :restart, "service[#{TEAMCITY_SERVICE_NAME}]", :delayed
end

service TEAMCITY_SERVICE_NAME do
  supports start: true, stop: true, restart: true, status: true
  action [:enable, :start]
end


recipes/windows_agent.rb
#
# Cookbook Name:: chef-teamcity
# Recipe:: agent-windows
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

TEAMCITY_VERSION = node['teamcity']['version'].freeze
TEAMCITY_SERVICE_NAME = 'TCBuildAgent'.freeze

TEAMCITY_AGENT_NAME = node['teamcity']['agent']['name'].freeze
TEAMCITY_AGENT_SERVER_URI = node['teamcity']['agent']['server_uri'].freeze
TEAMCITY_AGENT_FILE = 'buildAgent.zip'.freeze
TEAMCITY_AGENT_URI = ::URI.join(TEAMCITY_AGENT_SERVER_URI, "update/#{TEAMCITY_AGENT_FILE}").to_s.freeze
TEAMCITY_AGENT_WORK_DIR = node['teamcity']['agent']['work_dir'].freeze
TEAMCITY_AGENT_TEMP_DIR = node['teamcity']['agent']['temp_dir'].freeze
TEAMCITY_AGENT_SYSTEM_DIR = node['teamcity']['agent']['system_dir'].freeze
TEAMCITY_AGENT_OWN_ADDRESS = node['teamcity']['agent']['own_address'].freeze
TEAMCITY_AGENT_OWN_PORT = node['teamcity']['agent']['port'].freeze
TEAMCITY_AGENT_AUTH_TOKEN = node['teamcity']['agent']['authorization_token'].freeze
TEAMCITY_AGENT_SYSTEM_PROPERTIES = node['teamcity']['agent']['system_properties'].freeze
TEAMCITY_AGENT_ENV_PROPERTIES = node['teamcity']['agent']['env_properties'].freeze
TEAMCITY_AGENT_SRC_PATH = ::File.join(TEAMCITY_AGENT_SYSTEM_DIR, TEAMCITY_AGENT_FILE).freeze
TEAMCITY_AGENT_CONFIG_PATH = "#{TEAMCITY_AGENT_SYSTEM_DIR}/conf".freeze
TEAMCITY_AGENT_PROPERTIES = "#{TEAMCITY_AGENT_CONFIG_PATH}/buildAgent.properties".freeze
TEAMCITY_AGENT_BIN_PATH = ::File.join(TEAMCITY_AGENT_SYSTEM_DIR, 'bin')

TEAMCITY_SRC_PATH = "#{TEAMCITY_AGENT_SYSTEM_DIR}.zip".freeze
TEAMCITY_PID_FILE = "#{TEAMCITY_AGENT_SYSTEM_DIR}\\logs\\buildAgent.pid".freeze

remote_file TEAMCITY_SRC_PATH do
  source TEAMCITY_AGENT_URI
  not_if { ::File.exist?(TEAMCITY_AGENT_CONFIG_PATH) }
end

directory TEAMCITY_AGENT_SYSTEM_DIR do
  action :create
  recursive true
  not_if { ::File.exist?(TEAMCITY_AGENT_CONFIG_PATH) }
end

windows_zipfile TEAMCITY_AGENT_SYSTEM_DIR do
  source TEAMCITY_SRC_PATH
  action :unzip
  not_if { ::File.exist?(TEAMCITY_AGENT_BIN_PATH) }
end

template TEAMCITY_AGENT_PROPERTIES do
  source 'buildAgent.properties.erb'
  variables(
              server_uri: TEAMCITY_AGENT_SERVER_URI,
              name: TEAMCITY_AGENT_NAME,
              work_dir: TEAMCITY_AGENT_WORK_DIR,
              temp_dir: TEAMCITY_AGENT_TEMP_DIR,
              system_dir: TEAMCITY_AGENT_SYSTEM_DIR,
              own_address: TEAMCITY_AGENT_OWN_ADDRESS,
              own_port: TEAMCITY_AGENT_OWN_PORT,
              authorization_token: TEAMCITY_AGENT_AUTH_TOKEN,
              system_properties: TEAMCITY_AGENT_SYSTEM_PROPERTIES,
              env_properties: TEAMCITY_AGENT_ENV_PROPERTIES
            )
  not_if { ::File.exist?(TEAMCITY_AGENT_PROPERTIES) }
  notifies :restart, "service[#{TEAMCITY_SERVICE_NAME}]", :delayed
end

execute 'install teamcity service' do
  command "#{TEAMCITY_AGENT_BIN_PATH}/service.install.bat"
  action :run
  cwd "#{TEAMCITY_AGENT_SYSTEM_DIR}/bin"
  not_if { ::Win32::Service.exists?('TCBuildAgent') }
end

service TEAMCITY_SERVICE_NAME do
  supports start: true, stop: true, restart: true, status: true
  action [:enable, :start]
end


recipes/agent-linux.rb
#
# Cookbook Name:: chef-teamcity
# Recipe:: agent-linux
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

TEAMCITY_VERSION = node['teamcity']['version'].freeze
TEAMCITY_USERNAME = node['teamcity']['username'].freeze
TEAMCITY_SERVICE_NAME = node['teamcity']['service_name'].freeze
TEAMCITY_GROUP = node['teamcity']['group'].freeze
TEAMCITY_HOME_PATH = "/home/#{TEAMCITY_USERNAME}".freeze
TEAMCITY_PATH = "/opt/TeamCity-#{TEAMCITY_VERSION}".freeze
TEAMCITY_INIT_LOCATION = "/etc/init.d/#{TEAMCITY_SERVICE_NAME}".freeze
TEAMCITY_EXECUTABLE_MODE = 0755
TEAMCITY_READ_MODE = 0644

TEAMCITY_SRC_PATH = "#{TEAMCITY_PATH}.zip".freeze
TEAMCITY_PID_FILE = "#{TEAMCITY_PATH}/logs/buildAgent.pid".freeze
TEAMCITY_AGENT_NAME = node['teamcity']['agent']['name'].freeze
TEAMCITY_AGENT_SERVER_URI = node['teamcity']['agent']['server_uri'].freeze
TEAMCITY_AGENT_FILE = 'buildAgent.zip'.freeze
TEAMCITY_AGENT_URI = ::URI.join(TEAMCITY_AGENT_SERVER_URI, "update/#{TEAMCITY_AGENT_FILE}").to_s.freeze
TEAMCITY_AGENT_SRC_PATH = ::File.join(TEAMCITY_PATH, TEAMCITY_AGENT_FILE).freeze
TEAMCITY_AGENT_CONFIG_PATH = "#{TEAMCITY_PATH}/conf".freeze
TEAMCITY_AGENT_PROPERTIES = "#{TEAMCITY_AGENT_CONFIG_PATH}/buildAgent.properties".freeze
TEAMCITY_AGENT_EXECUTABLE = "#{TEAMCITY_PATH}/bin/agent.sh".freeze
TEAMCITY_AGENT_WORK_DIR = node['teamcity']['agent']['work_dir'].freeze
TEAMCITY_AGENT_TEMP_DIR = node['teamcity']['agent']['temp_dir'].freeze
TEAMCITY_AGENT_SYSTEM_DIR = node['teamcity']['agent']['system_dir'].freeze
TEAMCITY_AGENT_OWN_ADDRESS = node['teamcity']['agent']['own_address'].freeze
TEAMCITY_AGENT_OWN_PORT = node['teamcity']['agent']['port'].freeze
TEAMCITY_AGENT_AUTH_TOKEN = node['teamcity']['agent']['authorization_token'].freeze
TEAMCITY_AGENT_SYSTEM_PROPERTIES = node['teamcity']['agent']['system_properties'].freeze
TEAMCITY_AGENT_ENV_PROPERTIES = node['teamcity']['agent']['env_properties'].freeze

package 'unzip'

remote_file TEAMCITY_SRC_PATH do
  source TEAMCITY_AGENT_URI
  owner TEAMCITY_USERNAME
  group TEAMCITY_GROUP
  mode TEAMCITY_READ_MODE
  not_if { ::File.exists?(TEAMCITY_PATH) }
end

bash 'extract_teamcity' do
  cwd '/opt'
  code <<-EOH
    unzip #{TEAMCITY_SRC_PATH} -d #{TEAMCITY_PATH}
    chown -R #{TEAMCITY_USERNAME}.#{TEAMCITY_GROUP} #{TEAMCITY_PATH}
    rm -f #{TEAMCITY_SRC_PATH}
  EOH
  not_if { ::File.exists?(TEAMCITY_PATH) }
end

file TEAMCITY_AGENT_EXECUTABLE do
  owner TEAMCITY_USERNAME
  group TEAMCITY_GROUP
  mode TEAMCITY_EXECUTABLE_MODE
end

[TEAMCITY_AGENT_CONFIG_PATH].each do |p|
  directory p do
    owner TEAMCITY_USERNAME
    group TEAMCITY_GROUP
    recursive true
    mode TEAMCITY_EXECUTABLE_MODE
  end
end

template TEAMCITY_AGENT_PROPERTIES do
  source 'buildAgent.properties.erb'
  mode TEAMCITY_READ_MODE
  owner TEAMCITY_USERNAME
  group TEAMCITY_GROUP
  variables({
              server_uri: TEAMCITY_AGENT_SERVER_URI,
              name: TEAMCITY_AGENT_NAME,
              work_dir: TEAMCITY_AGENT_WORK_DIR,
              temp_dir: TEAMCITY_AGENT_TEMP_DIR,
              system_dir: TEAMCITY_AGENT_SYSTEM_DIR,
              own_address: TEAMCITY_AGENT_OWN_ADDRESS,
              own_port: TEAMCITY_AGENT_OWN_PORT,
              authorization_token: TEAMCITY_AGENT_AUTH_TOKEN,
              system_properties: TEAMCITY_AGENT_SYSTEM_PROPERTIES,
              env_properties: TEAMCITY_AGENT_ENV_PROPERTIES
            })
  not_if { ::File.exists?(TEAMCITY_AGENT_PROPERTIES) }
  notifies :restart, "service[#{TEAMCITY_SERVICE_NAME}]", :delayed
end

template TEAMCITY_INIT_LOCATION do
  source 'teamcity_agent_init.erb'
  mode TEAMCITY_EXECUTABLE_MODE
  owner 'root'
  group 'root'
  variables({
              teamcity_user_name: TEAMCITY_USERNAME,
              teamcity_executable: TEAMCITY_AGENT_EXECUTABLE,
              teamcity_pidfile: TEAMCITY_PID_FILE,
              teamcity_service_name: TEAMCITY_SERVICE_NAME
            })
  notifies :restart, "service[#{TEAMCITY_SERVICE_NAME}]", :delayed
end

service TEAMCITY_SERVICE_NAME do
  supports start: true, stop: true, restart: true, status: true
  action [:enable, :start]
end
